sshd/zenith/sshd/backends/crd.py

```diff
@@ -0,0 +1,182 @@
+import datetime
+import logging
+import typing as t
+
+from easykube import Configuration, ApiError
+
+from .. import config
+
+from . import base
+
+
+def isotime() -> str:
+    """
+    Returns the current time as an ISO8601 formatted string.
+    """
+    return datetime.datetime.now(tz = datetime.timezone.utc).isoformat(timespec = "seconds")
+
+
+class Backend(base.Backend):
+    """
+    SSHD backend that stores information using CRD instances.
+    """
+    def __init__(
+        self,
+        logger: logging.Logger,
+        api_version: str,
+        target_namespace: str
+    ):
+        self.logger = logger
+        self.api_version = api_version
+        # Initialise an easykube client from the environment
+        self.ekclient = Configuration.from_environment().sync_client(
+            default_namespace = target_namespace
+        )
+
+    def tunnel_check_host_and_port(self, host: str, port: int) -> bool:
+        ekendpoints = self.ekclient.api(self.api_version).resource("endpoints")
+        for endpoints in ekendpoints.list():
+            for endpoint in endpoints.get("spec", {}).get("endpoints", {}).values():
+                if endpoint["address"] == host and endpoint["port"] == port:
+                    return False
+        return True
+
+    def tunnel_init(
+        self,
+        subdomain: str,
+        host: str,
+        port: int,
+        ttl: int,
+        reap_after: int,
+        config_dict: t.Dict[str, t.Any]
+    ) -> str:
+        # Get the endpoints record, so that it can own our lease
+        # If it doesn't exist, create it
+        ekservices = self.ekclient.api(self.api_version).resource("services")
+        ekendpoints = self.ekclient.api(self.api_version).resource("endpoints")
+        try:
+            endpoints = ekendpoints.fetch(subdomain)
+        except ApiError as exc:
+            if exc.status_code == 404:
+                # Fetch the service so it can own the endpoints resource
+                service = ekservices.fetch(subdomain)
```

Comment:
I think you should fetch this first, and error on 404, to stop any surprises? I know its slightly less efficient, but I think it will be simpler to reason about.